src/ucs/debug/log.c

```diff
@@ -94,32 +106,124 @@ void ucs_log_flush()
                           ucs_log_level_t level,
                           const ucs_log_component_config_t *comp_conf,
                           const struct timeval *tv, const char *message)
 {
-    char *valg_buf;
+    const char *next_token;
+    char old_log_file_name[UCS_LOG_MAX_FILE_NAME];
+    char new_log_file_name[UCS_LOG_MAX_FILE_NAME];
+    char *log_buf;
+    int log_entry_len, ret;
+    unsigned idx, start_idx;
 
     if (RUNNING_ON_VALGRIND) {
-        valg_buf = ucs_alloca(buffer_size + 1);
-        snprintf(valg_buf, buffer_size,
-                 "[%lu.%06lu] %16s:%-4u %-4s %-5s %s\n", tv->tv_sec, tv->tv_usec,
+        log_buf = ucs_alloca(buffer_size + 1);
+        snprintf(log_buf, buffer_size, UCS_LOG_SHORT_FMT,
+                 tv->tv_sec, tv->tv_usec,
                  short_file, line, comp_conf->name,
                  ucs_log_level_names[level], message);
-        VALGRIND_PRINTF("%s", valg_buf);
+        VALGRIND_PRINTF("%s", log_buf);
     } else if (ucs_log_initialized) {
-        fprintf(ucs_log_file,
-                "[%lu.%06lu] [%s:%-5d:%d] %16s:%-4u %-4s %-5s %s\n",
+        if (ucs_log_file_close) { /* non-stdout/stderr */
+            /* get log entry size */
+            log_entry_len = snprintf(NULL, 0, UCS_LOG_FMT,
+                                     tv->tv_sec, tv->tv_usec, ucs_log_hostname,
+                                     ucs_log_pid, ucs_log_get_thread_num(),
+                                     short_file, line, comp_conf->name,
+                                     ucs_log_level_names[level], message);
+            /* check if it is necessary to find a new storage for logs */
+            if ((log_entry_len +
+                 ftell(ucs_log_file)) >= ucs_global_opts.log_file_max_size) {
+                fclose(ucs_log_file);
+
+                if (ucs_global_opts.log_file_backup_count != 0) {
+                    if (ucs_log_file_last_idx == ucs_global_opts.log_file_backup_count) {
+                        /* remove the last file and log rotation from the
+                         * `log_file_backup_count - 1` file */
+                        ucs_log_get_file_name(old_log_file_name,
+                                              UCS_LOG_MAX_FILE_NAME,
+                                              ucs_log_file_last_idx);
+                        unlink(old_log_file_name);
+                        start_idx = ucs_log_file_last_idx - 1;
+                    } else {
+                        start_idx = ucs_log_file_last_idx;
+                        ucs_log_file_last_idx++;
+                    }
+
+                    ucs_assert(ucs_log_file_last_idx <=
+                               ucs_global_opts.log_file_backup_count);
+                    ucs_assert((start_idx <=
+                                (ucs_global_opts.log_file_backup_count - 1)) &&
+                               (start_idx != UINT_MAX));
+
+                    for (idx = start_idx; idx != UINT_MAX; --idx) {
```

Comment:
maybe defining `idx` and `start_idx`  as `ints` would eliminate working with UINT_MAX (which is a bit confusing)